course_info/tests/jasmine/controllers/DetailsControllerSpec.js

```diff
@@ -559,4 +566,161 @@ describe('Unit testing DetailsController', function () {
+            aEvent.keyCode = 13;
+            $("input").trigger(aEvent);
+            dc.associateNewSiteHandleKey(aEvent);
+            expect(dc.validNewSiteURL).toHaveBeenCalled();
+            expect(dc.associateNewSite).not.toHaveBeenCalled();
+        });
+
+    });
+
+    describe('associateNewSite', function(){
+        beforeEach(function () {
+            var ci = {
+                course_instance_id: $routeParams.courseInstanceId,
+                title: 'Test Course Title',
+                short_title: 'Test',
+                sub_title: 'Jasime is your friend',
+                description: '<p>hello</p>',
+                sync_to_canvas : 1,
+                exclude_from_isites:0,
+                exclude_from_catalog:0,
+                sites: [
+                    {
+                        external_id: 'https://x.y.z/888',
+                        site_id: '888',
+                        map_type: 'official'
+                    },
+                    {
+                        external_id: 'https://x.y.z/999',
+                        site_id: '999',
+                        map_type: 'unofficial'
+                    }
+                ]
+            };
+            dc = $controller('DetailsController', {$scope: scope});
+            dc.courseInstance = ci;
+            $httpBackend.expectGET(courseInstanceURL).respond(200, JSON.stringify({status: "success"}));
+            $httpBackend.expectGET(peopleURL).respond(200, JSON.stringify({status: "success"}));
+            $httpBackend.flush(2);
+        });
+
+        it('should make the post call to associate a new site with the course instance', function(){
+            dc.newCourseSiteURL = 'http://testing.com';
+            dc.associateNewSite();
+            $httpBackend.expectPOST(postSiteUrl)
+                .respond(201, JSON.stringify({status: "success"}));
+            $httpBackend.flush(1);
+            siteList = [{external_id: 'https://x.y.z/888', site_id: '888', map_type: 'official'},
+                        {external_id: 'https://x.y.z/999', site_id: '999', map_type: 'unofficial'},
+                        {course_site_url: 'http://testing.com', map_type: 'official', site_map_id: undefined}
+                        ];
+            expect(dc.courseInstance.sites.length).toEqual(3);
+            expect(dc.courseInstance.sites).toEqual(siteList);
+        });
+
+        it('should show an error message if the post fails', function(){
+            dc.newCourseSiteURL = 'http://testing.com';
+            dc.associateNewSite();
+            $httpBackend.expectPOST(postSiteUrl)
+                .respond(500, JSON.stringify({status: "invalid url"}));
+            $httpBackend.flush(1);
+            expect(dc.alerts.form.siteOperationFailed).toEqual({show: true, operation: 'associating', details: 'None'});
+        });
+
+    });
+
+    describe('dissociateSite', function(){
+        beforeEach(function () {
+            var ci = {
```

Comment:
Seems like much of this could be DRYer as well -- but I recognize some people prefer the test data as part of the main `describe()` block, even if there's lots of duplication.